MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java

```diff
@@ -0,0 +1,175 @@
+package com.google.moviestvsentiments.service.web;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
+import androidx.lifecycle.LiveData;
+import androidx.lifecycle.MutableLiveData;
+import com.google.moviestvsentiments.util.LiveDataTestUtil;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import junitparams.JUnitParamsRunner;
+import junitparams.Parameters;
+import retrofit2.Response;
+
+@RunWith(JUnitParamsRunner.class)
+public class NetworkBoundResourceTest {
+
+    private static final String LOCAL_VALUE = "Local Value";
+    private static final String SERVER_VALUE = "Server Value";
+
+    @Rule
+    public InstantTaskExecutorRule instantExecutorRule = new InstantTaskExecutorRule();
+
+    private Object[] invokesCorrectMethodsParameters() {
+        return new Object[] {
+            new Object[] {false, new boolean[] {false, true, true, false} },
+            new Object[] {true, new boolean[] {true, true, true, true} }
+        };
+    }
+
+    @Test
+    @Parameters(method = "invokesCorrectMethodsParameters")
+    public void networkBoundResource_invokesCorrectMethods(boolean shouldFetchReturnValue,
+                                                           boolean[] expectedInvocations) {
+        final boolean[] invocations = {false, false, false, false};
+        NetworkBoundResource resource = new NetworkBoundResource<String, String>() {
+            @Override protected void saveCallResult(String item) {
+                invocations[0] = true;
+            }
+
+            @Override protected boolean shouldFetch(String data) {
+                invocations[1] = true;
+                return shouldFetchReturnValue;
+            }
+
+            @Override protected LiveData<String> loadFromRoom() {
+                invocations[2] = true;
+                return new MutableLiveData<>("testValue");
+            }
+
+            @Override protected LiveData<ApiResponse<String>> performNetworkCall() {
+                invocations[3] = true;
+                return new MutableLiveData<>(new ApiResponse(Response.success("Server value")));
+            }
+        };
+
+        resource.getResult().observeForever(data -> {});
+
+        assertThat(invocations).isEqualTo(expectedInvocations);
+    }
+
+    @Test
+    public void networkBoundResource_passesCorrectParameters() {
+        final String[] shouldFetchParameter = new String[1];
+        final String[] saveCallResultParameter = new String[1];
```

Comment:
Why not use String?

Reply:
That results in the error: `Variable 'shouldFetchParameter' is accessed from within inner class, needs to be final or effectively final`

Reply:
I see. thanks